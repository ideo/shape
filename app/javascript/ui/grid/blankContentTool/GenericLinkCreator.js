import PropTypes from 'prop-types'

import Button from '~/ui/global/Button'
import { BctTextField } from '~/ui/global/styled/forms'
import PaddedCardCover from '~/ui/grid/covers/PaddedCardCover'
import { KEYS } from '~/utils/variables'
import ValidIndicator from './ValidIndicator'

// NOTE: This component used to be shared between LinkCreator and VideoCreator
// but now those have merged into one.
const GenericLinkCreator = ({
  url,
  urlValid,
  loading,
  placeholder,
  onSubmit,
  onChange,
  onClose,
  password,
  passwordField,
  onPasswordChange,
}) => {
  let validIndicator, pwValidIndicator
  if (url.length > 3) {
    validIndicator = <ValidIndicator valid={!!urlValid} loading={loading} />
  }

  if (passwordField) {
    pwValidIndicator = (
      <ValidIndicator top={58} valid={false} loading={loading} />
    )
  }

  const handleKeyDown = e => {
    if (e.keyCode === KEYS.ESC) {
      onClose()
    }
  }

  return (
    <PaddedCardCover>
      <form className="form" onSubmit={onSubmit}>
        <BctTextField
          data-cy="BctTextField"
          autoFocus
          placeholder={placeholder}
          value={url}
          onChange={onChange}
          onKeyDown={handleKeyDown}
        />
        {passwordField && (
          <BctTextField
            type="password"
            autoFocus
            placeholder="Enter password"
            value={password}
            onChange={onPasswordChange}
          />
        )}
        {validIndicator}
        {pwValidIndicator}
        <Button disabled={loading} data-cy="LinkCreatorFormButton">
          Add
        </Button>
      </form>
    </PaddedCardCover>
  )
}

GenericLinkCreator.propTypes = {
  url: PropTypes.string.isRequired,
  urlValid: PropTypes.bool.isRequired,
  password: PropTypes.string,
  passwordField: PropTypes.bool,
  onPasswordChange: PropTypes.func,
  placeholder: PropTypes.string.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired,
}

GenericLinkCreator.defaultProps = {
  password: '',
  passwordField: false,
  onPasswordChange: null,
}

export default GenericLinkCreator
