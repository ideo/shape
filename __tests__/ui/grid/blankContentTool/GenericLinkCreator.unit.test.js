import GenericLinkCreator from '~/ui/grid/blankContentTool/GenericLinkCreator'
import expectTreeToMatchSnapshot from '#/helpers/expectTreeToMatchSnapshot'

let wrapper, props
describe('GenericLinkCreator', () => {
  beforeEach(() => {
    props = {
      url: '',
      urlValid: false,
      loading: false,
      placeholder: 'Create link',
      onSubmit: jest.fn(),
      onChange: jest.fn(),
      onClose: jest.fn(),
      password: '',
      passwordField: false,
      onPasswordChange: jest.fn(),
    }
    wrapper = shallow(<GenericLinkCreator {...props} />)
  })

  it('renders snapshot', () => {
    expectTreeToMatchSnapshot(wrapper)
  })

  it('renders a form with a BctTextField', () => {
    expect(wrapper.find('form').props().onSubmit).toEqual(props.onSubmit)
    expect(wrapper.find('BctTextField').length).toEqual(1)
    expect(wrapper.find('BctTextField').get(0).props.placeholder).toEqual(
      props.placeholder
    )
  })

  it('calls onChange callback', () => {
    const textField = wrapper.find('BctTextField').get(0)
    expect(textField.props.onChange).toEqual(props.onChange)
  })

  describe('with passwordField', () => {
    beforeEach(() => {
      wrapper = shallow(<GenericLinkCreator {...props} passwordField />)
    })

    it('renders a passwordField if indicated', () => {
      expect(wrapper.find('BctTextField').length).toEqual(2)
      const pwField = wrapper.find('BctTextField').get(1)
      expect(pwField.props.type).toEqual('password')
      expect(pwField.props.placeholder).toEqual('Enter password')
    })

    it('calls onPasswordChange callback', () => {
      const pwField = wrapper.find('BctTextField').get(1)
      expect(pwField.props.onChange).toEqual(props.onPasswordChange)
    })
  })
})
