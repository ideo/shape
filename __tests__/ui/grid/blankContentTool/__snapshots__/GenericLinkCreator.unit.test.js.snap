// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericLinkCreator renders snapshot 1`] = `
<styled.div>
  <form
    className="form"
    onSubmit={[MockFunction]}
  >
    <BctTextField
      autoFocus={true}
      data-cy="BctTextField"
      onChange={[MockFunction]}
      onKeyDown={[Function]}
      placeholder="Create link"
      value=""
    />
    <StyledFormButton
      color="#120f0e"
      data-cy="LinkCreatorFormButton"
      disabled={false}
    >
      Add
    </StyledFormButton>
  </form>
</styled.div>
`;
